packages/@sanity/form-builder/src/inputs/arrays/ArrayOfPrimitivesInput/ArrayOfPrimitivesInput.tsx

```diff
@@ -184,10 +285,18 @@ export class ArrayOfPrimitivesInput extends React.PureComponent<Props> {
                     const childPresence = presence.filter((pItem) =>
                       startsWith([index], pItem.path)
                     )
+                    const memberType = this.getMemberType(resolveTypeName(item))
+
+                    // Best effort attempt to make a stable key for each item in the array
+                    // Since items may be reordered and change at any time, there's no way to reliably address each item uniquely
+                    // (hashing the item value would impact performance, and it's common to have duplicate values)
+                    // This is a "best effort"-attempt at making sure we don't re-use internal state for item inputs
+                    // when items gets added or removed to the array
 
+                    const key =
+                      (memberType?.name || 'invalid-type') + String(index) + String(value.length)
```

Comment:
I think this key may cause all items to re-render if a new item is added to the array. was that intended? I think a key with the member type name and the index is about the best you can do

edit: noticed another small thing with the existing key: two numbers concatenated could cause collisions in some rare edge cases. adding a separator would fix it: e.g. `${index}_${value.length}`

Reply:
> noticed another small thing with the existing key: two numbers concatenated could cause collisions in some rare edge cases
ooh, great catch. fixed!